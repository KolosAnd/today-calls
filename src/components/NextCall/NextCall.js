import React, {useEffect, useState} from "react";
import {Title} from "../Title/Title";
import styles from "./NextCall.module.css"
import {getTimeCall, translateTimeStringToMiliseconds} from "../../utils/timeFunc";
import {sortTimeAscending} from "../../utils/sortingFunc";

const NextCall = ({calls}) => {
    const [callsSortArr, setCallsSortArr] = useState([]);
    const [nextCall, setNextCall] = useState({});

    useEffect(  () => {
        if(calls.length > 0){
            setCallsSortArr(calls.sort((a,b) =>  sortTimeAscending(a,b)));
        }
    },[calls]);

    useEffect(()=> {

        if(callsSortArr.length > 0) {

            for(let i=0; i<callsSortArr.length; i++){
                const timeInSeconds = translateTimeStringToMiliseconds(callsSortArr[i].time);
                const timeInMilisec = getTimeCall(timeInSeconds, Date.now());

                if (Date.now() < timeInMilisec) {setNextCall(callsSortArr[i]); break;}
                else {setNextCall(null)}
            }
            const timerId = timeNextCall();
            return  (() =>{
                clearInterval(timerId);
            })
        }
    },[callsSortArr]);

    const timeNextCall = () => {

    return setInterval(() =>{
            for(let i=0; i<callsSortArr.length; i++){
                const timeInSeconds = translateTimeStringToMiliseconds(callsSortArr[i].time);
                const timeInMilisec = getTimeCall(timeInSeconds, Date.now());
                if (Date.now() < timeInMilisec) {setNextCall(callsSortArr[i]); break;}
            }
        },60000);

    }

    return (
        <div className={styles.nextCall}>
            <div className={styles.nextCall__wrap}>
                <Title text={"Next call"} />
                {nextCall ?
                    <div className={styles.nextCall__inputWrap}>
                        <div className={styles.nextCall__input + ' ' + styles.nextCall__inputName}>{nextCall.name}</div>
                        <div className={styles.nextCall__input  + ' ' + styles.nextCall__inputPhone}>{nextCall.phone}</div>
                        <div className={styles.nextCall__input  + ' ' + styles.nextCall__inputTime}> {nextCall.time}</div>
                    </div>
                    :
                    <div className={styles.nextCall__inputWrap}>
                        <div className={styles.nextCall__input + ' ' + styles.nextCall__inputName}>No next call</div>
                    </div>

                }

            </div>
        </div>
    )
}
export default NextCall;