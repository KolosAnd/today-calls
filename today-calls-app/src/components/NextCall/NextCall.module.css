.nextCall {
    width: 35%;
}

@media screen and (max-width: 1200px) {
    .nextCall {
        width: 40%;
    }
}

@media screen and (max-width: 710px) {
    .nextCall {
        width: 100%;
        margin-bottom: 35px;
    }
}

.nextCall__wrap {
    border-radius: 5px;
    border: 2px solid #fac863;
    padding: 30px 25px;
}

@media screen and (max-width: 500px) {
    .nextCall__wrap {
        padding: 20px 15px;
    }
}

.nextCall__title {
    color: #fff;
    margin: 0 0 25px;
    font-size: 25px;
    text-decoration: none;
    line-height: 25px;
}

.nextCall__inputWrap {
    width: 100%;
    display: flex;
    justify-content: space-between;
}

@media screen and (max-width: 1200px) {
    .nextCall__inputsWrap {
        flex-wrap: wrap;
    }
}

.nextCall__input {
    font-size: 16px;
    line-height: 18px;
    color: #fff;
}

@media screen and (max-width: 1024px) {
    .nextCall__input:not(:last-child) {
        margin-bottom: 10px;
    }
}

.nextCall__inputName {
    width: 35%;
}

@media screen and (max-width: 1024px) {
    .nextCall__inputName {
        width: 100%;
    }
}

.nextCall__inputPhone {
    width: 35%;
}

@media screen and (max-width: 1024px) {
    .nextCall__inputPhone {
        width: 100%;
    }
}

.nextCall__inputTime {
    width: 20%;
}

@media screen and (max-width: 1024px) {
    .nextCall__inputTime {
        width: 100%;
    }
}
