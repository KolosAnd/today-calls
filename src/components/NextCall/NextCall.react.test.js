import React from 'react';
import renderer from 'react-test-renderer';
import NextCall from "./NextCall.js";

const calls = [
    {
        id: 1641294575511,
        milisec: 1641273300511,
        name: "Петя",
        phone: "056-34-45-111",
        time: "07:15",
    },
    {
        id: 1641294575611,
        milisec: 1641273300511,
        name: "Vasia",
        phone: "056-34-45-111",
        time: "07:11"
    },
    {
        id: 1641294575521,
        milisec: 1641273300511,
        name: "Misha",
        phone: "056-34-45-111",
        time: "09:25"
    },
    {
        id: 1641294574444,
        milisec: 1641273302451,
        name: "Grisha",
        phone: "056-34-45-111",
        time: "11:17"
    }

];

const callsFake = [];


const callsOneUndefined = [
    {
        id: undefined,
        milisec: undefined,
        name: undefined,
        phone: undefined,
        time: undefined,
    },
];

const callsOneNull = [
    {
        id: null,
        milisec: null,
        name: null,
        phone: null,
        time: null,
    },
];

describe('NextCall component', () => {

    test('NextCall with props', () => {
        const component = renderer.create(<NextCall calls={calls}/>,);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('NextCall with props fake calls', () => {
        const component = renderer.create(<NextCall calls={callsFake}/>,);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('NextCall with props one call undefined', () => {
        const component = renderer.create(<NextCall calls={callsOneUndefined}/>,);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('NextCall with props one call null', () => {
        const component = renderer.create(<NextCall calls={callsOneNull}/>,);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

})