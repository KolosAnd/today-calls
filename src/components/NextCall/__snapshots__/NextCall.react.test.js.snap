// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NextCall component NextCall with props 1`] = `
<div
  className="nextCall"
>
  <div
    className="nextCall__wrap"
  >
    <div
      className="title"
    >
      <h2
        className="title__name undefined"
      >
        Next call
      </h2>
    </div>
    <div
      className="nextCall__inputWrap"
    >
      <div
        className="nextCall__input nextCall__inputName"
      />
      <div
        className="nextCall__input nextCall__inputPhone"
      />
      <div
        className="nextCall__input nextCall__inputTime"
      >
         
      </div>
    </div>
  </div>
</div>
`;

exports[`NextCall component NextCall with props fake calls 1`] = `
<div
  className="nextCall"
>
  <div
    className="nextCall__wrap"
  >
    <div
      className="title"
    >
      <h2
        className="title__name undefined"
      >
        Next call
      </h2>
    </div>
    <div
      className="nextCall__inputWrap"
    >
      <div
        className="nextCall__input nextCall__inputName"
      />
      <div
        className="nextCall__input nextCall__inputPhone"
      />
      <div
        className="nextCall__input nextCall__inputTime"
      >
         
      </div>
    </div>
  </div>
</div>
`;

exports[`NextCall component NextCall with props one call undefined 1`] = `
<div
  className="nextCall"
>
  <div
    className="nextCall__wrap"
  >
    <div
      className="title"
    >
      <h2
        className="title__name undefined"
      >
        Next call
      </h2>
    </div>
    <div
      className="nextCall__inputWrap"
    >
      <div
        className="nextCall__input nextCall__inputName"
      />
      <div
        className="nextCall__input nextCall__inputPhone"
      />
      <div
        className="nextCall__input nextCall__inputTime"
      >
         
      </div>
    </div>
  </div>
</div>
`;

exports[`NextCall component NextCall without props 1`] = `
<div
  className="nextCall"
>
  <div
    className="nextCall__wrap"
  >
    <div
      className="title"
    >
      <h2
        className="title__name undefined"
      >
        Next call
      </h2>
    </div>
    <div
      className="nextCall__inputWrap"
    >
      <div
        className="nextCall__input nextCall__inputName"
      />
      <div
        className="nextCall__input nextCall__inputPhone"
      />
      <div
        className="nextCall__input nextCall__inputTime"
      >
         
      </div>
    </div>
  </div>
</div>
`;
